lib/CoreDisTools/coredistools.cpp

```diff
@@ -0,0 +1,236 @@
+  }
+
+  string Mcpu;        // Not specifying any particular CPU type.
+  string FeaturesStr; // No additional target specific attributes.
+  STI.reset(TheTarget->createMCSubtargetInfo(TargetTriple, Mcpu, FeaturesStr));
+  if (!STI) {
+    errs() << "error: no subtarget info for target " << TargetTriple << "\n";
+    return false;
+  }
+
+  MII.reset(TheTarget->createMCInstrInfo());
+  if (!MII) {
+    errs() << "error: no instruction info for target " << TargetTriple << "\n";
+    return false;
+  }
+
+  MOFI.reset(new MCObjectFileInfo);
+  Ctx.reset(new MCContext(AsmInfo.get(), MRI.get(), MOFI.get()));
+
+  Disassembler.reset(TheTarget->createMCDisassembler(*STI, *Ctx));
+
+  if (!Disassembler) {
+    errs() << "error: no disassembler for target " << TargetTriple << "\n";
+    return false;
+  }
+
+  int AsmPrinterVariant = AsmInfo->getAssemblerDialect();
+  IP.reset(TheTarget->createMCInstPrinter(
+      Triple(TargetTriple), AsmPrinterVariant, *AsmInfo, *MII, *MRI));
+
+  if (!IP) {
+    errs() << "error: No Instruction Printer for target " << TargetTriple
+           << "\n";
+    return false;
+  }
+
+  return true;
+}
+
+size_t CorDisasm::disasmInstruction(size_t Address, const uint8_t *Bytes,
+                                    size_t Maxlength, bool PrintAsm) const {
+  uint64_t Size;
+  MCInst Inst;
+  raw_ostream &CommentStream = nulls();
+  raw_ostream &DebugOut = nulls();
+  ArrayRef<uint8_t> ByteArray(Bytes, Maxlength);
+
+  bool success = Disassembler->getInstruction(Inst, Size, ByteArray, Address,
+                                              DebugOut, CommentStream);
+
+  if (!success) {
+    errs() << "Invalid instruction encoding\n";
+    return 0;
+  }
+
+  if (PrintAsm) {
+    printInstruction(&Inst, Address, Size, ByteArray.slice(0, Size));
+  }
+
+  return Size;
+}
+
+void CorDisasm::printInstruction(const MCInst *MI, size_t Address,
+                                 size_t InstSize,
+                                 ArrayRef<uint8_t> Bytes) const {
+  outs() << format("%8" PRIx64 ":", Address);
+  if (ShowInstructionBytes) {
+    outs() << "\t";
+    dumpBytes(Bytes.slice(0, InstSize), outs());
+  }
+  IP->printInst(MI, outs(), "", *STI);
+  outs() << "\n";
+}
+
+// Allocate and initialize a Disassembler object.
+// Returns the disassembler on success, nullptr on failure.
+DllIface CorDisasm *InitDisasm(TargetArch Target) {
+
+  CorDisasm *Disassembler = new CorDisasm(Target);
+  if (Disassembler->init()) {
+    return Disassembler;
+  }
+
+  delete Disassembler;
+  return nullptr;
+}
+
+DllIface void FinishDisasm(const CorDisasm *Disasm) { delete Disasm; }
+
+DllIface size_t DisasmInstruction(const CorDisasm *Disasm, size_t Address,
+                                  const uint8_t *Bytes, size_t Maxlength) {
+  assert((Disasm != nullptr) && "Disassembler object Expected ");
+  return Disasm->disasmInstruction(Address, Bytes, Maxlength);
+}
+
+DllIface void PrintInstruction(const CorDisasm *Disasm, size_t Address,
```

Comment:
We cannot use Optional arguments because of C compatibility. But I can add a mandatory argument to DisasmInstruction like: 

```
  DllIface size_t DisasmInstruction(const CorDisasm *Disasm, size_t Address,
                                                          const uint8_t *Bytes, size_t Maxlength, bool PrintAsm)
```

However I left it as two functions, because `DisasmInstruction` is designed for GcStress which is the sole user currently (to get the instruction size) and `PrintInstruction` is mostly used for debugging.